### ISSUE TITLE
Speed up constraint reflection

### ISSUE THREAD

**Issue by dmos62** [centerofci/mathesar#2662]
Refer to this comment for profile at the time of writing: https://github.com/centerofci/mathesar/issues/2544#issuecomment-1450451207

Currently, in `mathesar/state/django.py::reflect_constraints_from_database` we're recreating each constraint and overwriting it in the Django db. That's taking ~2.63/4.40s in the above mentioned profile. It seems to be the bulk create call and model initiations that seem to take up most of that time. Explore improving performance by being selective about which constraint models we instantiate and create.

**Comment by ShailRT** [comment]
Hi! can i work on this issue.

**Comment by rajatvijay** [comment]
@ShailRT sure, assigning this to you.

**Comment by ShailRT** [comment]
Are there any specific parameters for selection?
My first approach is to remove duplicate constraints.

**Comment by Aritra8438** [comment]
> we're recreating each constraint and overwriting it in the Django db

When are we recreating constraints?
The ```mathesar/state/django.py::reflect_constraints_from_database``` seems to be called, only when we are creating the table and adding a constraint. 

> It seems to be the bulk create call and model initiations that seem to take up most of that time

Does this happen only when we create a table by importing a file?

**Comment by dmos62** [comment]
@ShailRT the people that know this code the best are quite busy for the next few weeks. You might have to figure out a lot of this yourself.

**Comment by rajatvijay** [comment]
Hi @ShailRT sorry for the hasty assignment. 

Please check point no 3 in our [contributing guide](https://wiki.mathesar.org/en/community/contributing#contributing-code). During the GSoC phase, we would like to follow this rule rather more strictly to encourage more contribution. 

That being said, please continue your work on this issue. Do not worry too much about the duplicate PRs situation. Even a duplicate PR can showcase your various skills for the purposes of GSoC proposal selection, and if you choose to work on another issue after that, you'll probably get assigned to it from the start.

**Comment by ShailRT** [comment]
Hello @rajatvijay sorry if I was wrong, but I followed the documentation before asking for assignment. Please correct me if I am wrong. Thanks

**Comment by ShailRT** [comment]
Hi! @dmos62 can you tell if there is any package installed to create runtime profile in this project. I would like to test my changes, Thanks

**Comment by dmos62** [comment]
@ShailRT I used `cProfile` and `snakeviz` for profiling. Here's a snippet I have saved in my notes:

```
import cProfile
import time

  with cProfile.Profile() as pr:
    some_func_to_profile()
  profile_file_name = f'{int(time.time())}.prof'
  pr.dump_stats(profile_file_name)
```

Consider that reflection will be affected by the datasets you have in your database. Make sure you're not profiling with an empty or unrealistic database.

**Comment by ShailRT** [comment]
@dmos62 Thanks! Really needed this.

**Comment by siddhart1o1** [comment]
@dmos62
I have made some changes to reflect_constraints_from_database() to optimize its performance by reducing the number of model instantiations and database writes. The updated function retrieves a set of existing constraint_oid values from the Django database and only creates new models.Constraint objects for constraints whose oid values are not already present in the set.

Although this change may not significantly increase performance, it should improve performance somewhat. let me know your thoughts, Here is the updated code:
```
def reflect_constraints_from_database(database):
    engine = create_mathesar_engine(database.name)
    db_constraints = get_constraints_with_oids(engine)
    map_of_table_oid_to_constraint_oids = defaultdict(list)
    for db_constraint in db_constraints:
        table_oid = db_constraint['conrelid']
        constraint_oid = db_constraint['oid']
        map_of_table_oid_to_constraint_oids[table_oid].append(constraint_oid)

    table_oids = map_of_table_oid_to_constraint_oids.keys()
    tables = models.Table.current_objects.filter(oid__in=table_oids)
    
    existing_constraint_oids = set(models.Constraint.current_objects.values_list('oid', flat=True))
    print(existing_constraint_oids)
    constraint_objs_to_create = []
    for table in tables:
        constraint_oids = map_of_table_oid_to_constraint_oids.get(table.oid, [])
        for constraint_oid in constraint_oids:
            if constraint_oid not in existing_constraint_oids:
                constraint_obj = models.Constraint(oid=constraint_oid, table=table)
                constraint_objs_to_create.append(constraint_obj)
    
    models.Constraint.current_objects.bulk_create(constraint_objs_to_create, ignore_conflicts=True)
    _delete_stale_dj_constraints(db_constraints, database)
    engine.dispose()
```

**Comment by dmos62** [comment]
@siddhart1o1 please submit this as a draft PR. That will make it easy for me what exactly the changes are. Also, perform profiling and put it in that PR thread; we can't fully evaluate performance improvements without measuring them.

**Comment by Aritra8438** [comment]
Hi, @dmos62, can I work on this issue in parallel?

**Comment by dmos62** [comment]
@Aritra8438 yes.

**Comment by dmos62** [comment]
Not relevant at the moment. Remove SQLAlchemy projects are currently ongoing. If optimizations in reflection are still needed, we'll likely revisit them after that. Link to first of Remove SQLAlchemy projects: https://github.com/centerofci/mathesar/issues/2737
